Streamline SSL experience

josevalim opened the pull request:
@chrismccord, let me know if this works. For now, this means you need to explicitly set `ssl: :verify_full` in your config file but we can support "ssl=verify_full" via query string too.

wojtekmach commented on lib/postgrex/protocol.ex:
```diff
@@ -117,14 +125,37 @@ defmodule Postgrex.Protocol do
+    case Keyword.fetch(opts, :hostname) do
+      {:ok, hostname} ->
+        [
+          server_name_indication: hostname,
```
```suggestion
          server_name_indication: String.to_charlist(hostname),
```

wojtekmach commented on lib/postgrex/protocol.ex:
```diff
@@ -117,14 +125,37 @@ defmodule Postgrex.Protocol do
+          server_name_indication: hostname,
+          customize_hostname_check: [
+            match_fun: :public_key.pkix_verify_hostname_match_fun(:https)
+          ]
```
Can we also add `verify: :verify_peer`? I believe it's not enabled by default.

```suggestion
          ],
          verify: :verify_peer
```

wojtekmach replied:
Sorry, please disregard this, I think it _is_ using verify_peer by default:

If I use just `ssl: :verify_full` I get this error:

```
11:13:24.657 [error] Postgrex.Protocol (#PID<0.203.0>) failed to connect: ** (DBConnection.ConnectionError) ssl connect: Options (or their values) can not be combined: [{verify,verify_peer},
                                                {cacerts,undefined}] - {:options, :incompatible, [verify: :verify_peer, cacerts: :undefined]}
```

Unfortunately I can't just pass `ssl_opts: [cacerts: :public_key.cacerts_get()]` because DO, like I'm sure most servers, are not using public certs? Passing:

```
ssl_opts: [
  cacertfile: "/Users/wojtek/Downloads/ca-certificate.crt"
]
```

that I got from them fixes the error.

wojtekmach reviewed (approved):
Just one comment about charlist and this is good to go. I tested it on DigitalOcean and it worked!

```elixir
Mix.install([
  {:postgrex, github: "elixir-ecto/postgrex", branch: "jv-add-ssl-verify-full"}
])

{:ok, pid} =
  Postgrex.start_link(
    hostname: System.fetch_env!("DO_HOST"),
    port: 25060,
    username: "doadmin",
    password: System.fetch_env!("DO_PASSWORD"),
    database: "defaultdb",
    ssl: :verify_full,
    ssl_opts: [
      cacertfile: "/Users/wojtek/Downloads/ca-certificate.crt"
    ]
  )

IO.inspect(Postgrex.query!(pid, "SELECT NOW()", []))
```

josevalim commented on lib/postgrex.ex:
```diff
@@ -118,7 +118,10 @@ defmodule Postgrex do
       Defaults to 1000ms;
 
-    * `:ssl` - Set to `true` if ssl should be used (default: `false`);
+    * `:ssl` - Set to `true` if ssl should be used (default: `false`). You may also
+      set it to `:verify_full`, which enables peer and hostname validation. You must
+      specify either `:cacerts` or `:cacertfile` in `:ssl_opts` if `:verify_full`
+      is enabled;
```
```suggestion
    * `:ssl` - Set to `:verify_full` to enable full SSL verification, including peer
      and hostname. You must also specify either `:cacerts` or `:cacertfile` in
      `:ssl_opts`. Alternatively, you may set it to `true` for enabling SSL, but
      use `:ssl_opts` to opt-in and control verification. Defaults to `false`.
```

josevalim commented:
@voltone, please let me know if you have thoughts here, the goal is to mirror this in myxql as well.

chrismccord reviewed (approved):
:shipit:

voltone commented on lib/postgrex/protocol.ex:
```diff
@@ -81,12 +81,29 @@ defmodule Postgrex.Protocol do
+        {true, opts} ->
+          Logger.warning(
+            "setting ssl: true on your database connection offers only limited protection, " <>
+              "as the hostname is not verified. Set \"ssl: [cacertfile: path/to/file]\" instead"
```
```suggestion
              "as the server's certificate is not verified. Set \"ssl: [cacertfile: path/to/file]\" instead"
```

voltone commented on lib/postgrex/protocol.ex:
```diff
@@ -117,14 +134,27 @@ defmodule Postgrex.Protocol do
   end
 
+  defp default_ssl_opts(opts) do
+    case Keyword.fetch(opts, :hostname) do
```
If hostname is not set and PGHOST is used instead, the `server_name_indication` won't reflect the actual hostname, and server name verification will fail (as the TLS handshake is started on an established TCP connection and `:ssl` does not know the hostname that was used to establish the connection in the first place).

So either the docs would have to state that `hostname` option is mandatory when using `verify_peer`, or the logic to set `server_name_indication` would have to move to a later place where the actual hostname is known, no?

josevalim replied:
I think this is good. PGHOST is read on `Postgrex.start_link` and set as the `opts[:hostname]`. If endpoints is set, that wins over hostname, but we will still use the hostname for SNI. I will update the docs!

voltone replied:
So if `hostname` is never supposed to be absent, why do we need the `case` statement?

josevalim replied:
@voltone do you know if :server_name_indicator has any effect if we have `verify: :verify_none`?

voltone reviewed (commented):
There are a few more places in the docs where ssl options are described, including for example the `endpoints` list under "Failover with SSL support". I don't have time right now to trace the behavior of `endpoints` and understand the implications to propose updates, sorry. Thought I'd give my feedback now before this gets merged. I might have time tonight...

voltone commented on lib/postgrex.ex:
```diff
@@ -118,11 +117,10 @@ defmodule Postgrex do
-    * `:ssl_opts` - A list of ssl options, see the
-      [`tls_client_option`](http://erlang.org/doc/man/ssl.html#type-tls_client_option)
-      from the ssl docs;
+    * `:ssl` - Enables SSL. Setting it to `true` enables SSL without host verification,
+      which emits a warning. Instead, prefer to set it to a keyword list, with either
+      `:cacerts` or `:cacertfile` pointing to the server certificate, to enable hostname
+      verification. Defaults to `false`;
```
Setting `verify_peer` enables server certificate verification, with hostname verification being just a part of that. Without it, the server certificate is completely ignored, not just the hostname.

Also, `:cacerts` and `:cacertfile` need to point to a CA trust store, not to the server certificate.

```suggestion
    * `:ssl` - Enables SSL. Setting it to `true` enables SSL without server certificate verification,
      which emits a warning. Instead, prefer to set it to a keyword list, with either
      `:cacerts` or `:cacertfile` set to a CA trust store, to enable server certificate
      verification. Defaults to `false`;
```

josevalim commented:
Thank you @voltone, I believe I have addressed all of your concerns. However, there is one problem, we are always setting `server_name_indicator`, even in the cases where we would previously set only `ssl: true` (i.e. when `verify: :verify_none`). Let me know if this can be a concern because then we can set `server_name_indicator: disabled` instead.

josevalim commented:
> However, there is one problem, we are always setting server_name_indicator, even in the cases where we would previously set only ssl: true (i.e. when verify: :verify_none)

From some tests this does not seem to be an issue and, if it happens, folks can always set it to `:disabled`.

wojtekmach reviewed (approved):
I tested this on Supabase, CockroachDB, DigitalOcean PostgreSQL, and PlanetScale (which is MySQL, but same idea) and it works well! I'll work on similar change for MyXQL soon.

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

ruslandoga commented:
👋

It seems like setting `?ssl=true` overwrites the keyword options at https://github.com/elixir-ecto/ecto/blob/cba494787b8e98aa0038f6c35e2f44d3661f2620/lib/ecto/repo/supervisor.ex#L28

```elixir
config :app, Repo,
  url: "ecto://postgresql://your_username:your_password@cluster-do-user-1234567-0.db.ondigitalocean.com:25060/defaultdb?ssl=true",
  ssl: [cacertfile: "/path/to/cacert.pem"]
```
results in
```elixir
Postgrex.child_spec([repo: Repo, telemetry_prefix: [:app, :repo], otp_app: :app, ...etc..., ssl: true, pool: DBConnection.ConnectionPool])
```
and a warning
```
[warning] setting ssl: true on your database connection offers only limited protection, as the server's certificate is not verified. Set "ssl: [cacertfile: path/to/file]" instead
```

I wonder if URL parsing logic should be updated to avoid overwriting `ssl` cacerts?

josevalim commented:
Yes, we should probably fix it in Ecto. Maybe we ignore `ssl: true` if there is already a list in place (and we emit a warning).

ruslandoga commented on lib/postgrex/protocol.ex:
```diff
@@ -81,12 +81,29 @@ defmodule Postgrex.Protocol do
+          )
+
+          # Read ssl_opts for backwards compatibility
+          Keyword.pop(opts, :ssl_opts, [])
```
👋 

Sorry for posting on an old PR, but should the default be `verify_peer: :none`? it seems like, by default, `:ssl` still tries to verify the peer.

```elixir
iex(1)> :ssl.connect(~c"google.com", 443, [])
{:error, {:options, :incompatible, [verify: :verify_peer, cacerts: :undefined]}}
```

wojtekmach replied:
I think previous OTP defaulted to :verify_none and more recent to :verify_peer and so I believe the idea behind this legacy `ssl: true` was to keep whatever that particular OTP defaulted to however `ssl: list` would force secure defaults.

wojtekmach replied:
Do you know if this change broke any apps? Cause if the warning implies less security we should change the warning not intentionally make things less secure. Sorry I don't have an easy way to check how it all works right now.

ruslandoga replied:
No, I don't think it broke any apps. My suggestion was purely a QoL improvement :)
